### PR TITLE
Refactor of 'carga' module, now renamed 'workload'

### PR DESCRIPTION
**Pull request by MathRPG**
Opening a _draft_ pull request for continuous review.
There is more refactoring to do and the code is in an unstable state (I've introduced bugs I'm yet to fix, and added classes I'm yet to document).
Any and all feedback is welcome.

### CODE REVIEW

**Comment by MathRPG** [comment]
Refactored 'carga' module, which was renamed to 'workload' in the process.
Main changes include:
* Added and updated documentation for virtually every new and existing methods and classes in the packages.
* A new hierarchy to separate concerns of "generating a workload" and "building a task"; the top-level interface "WorkloadGenerator" is fundamentally the public API of the package, but the inner abstract class "TaskBuilder" is for internal use only.
* Concrete implementations were moved to the subpackage "impl" .
* Workload generator serialization was moved away from .toString() methods, in favor of a more descriptive approach; toString methods were repurposed for a human-readable representation of the workload generators.
* Fixed bugs OP#83 and OP#84.

Some collateral changes include:
* LoadConfigurationWindow's spinners were changed to support real values were applicable. However, the backend still lacks certain support to process such values. Changes need to happen in workload serialization and iconic model interpretation.

Some directions to take in the future:
* Finish support for real-value distributions in global workload configurations.
* Introduce a Strategy pattern for task distribution among masters in a system (currently, only option is closest-to-fair)
* Rethink trace task parsing, since it is currently overly complex (and probably has bugs).
* Write unit tests to maintain and validate behavior.

File: src/ispd/arquivo/xml/utils/WrappedElement.java [hunk 1]
```diff
@@ -13,6 +14,49 @@
...
+    /**
+     * Construct a {@link TwoStageUniform} from this instance, with the values
+     * acquired from {@link #minimum()}, {@link #maximum()} and
+     * {@link #average()}.
+     *
+     * @return {@link TwoStageUniform} with a minimum, maximum, and average values
+     * as the respective tags in this instance, and {@code 0} for  probability.
+     */
+    public TwoStageUniform toTaskSizeNoProbability() {
+        return new TwoStageUniform(this.minimum(), this.average(), this.maximum(), 0);
```
Comment (MathRPG) [inline src/ispd/arquivo/xml/utils/WrappedElement.java:26]: Potentially wrong, this makes a "two stage uniform" that only picks values from one of the intervals.

File: src/ispd/arquivo/xml/utils/WrappedElement.java [hunk 2]
```diff
@@ -13,6 +14,49 @@
...
+     * Construct a {@link TwoStageUniform} from this instance, with the values
+     * acquired from {@link #minimum()} and {@link #maximum()}.
+     *
+     * @return {@link TwoStageUniform} with a minimum and maximum values as the
+     * respective tags in this instance, and default values for the fields
+     * {@link TwoStageUniform#intervalSplit()} and
+     * {@link TwoStageUniform#firstIntervalProbability()}.
+     * @see TwoStageUniform#TwoStageUniform(double, double)
+     */
+    public TwoStageUniform toTaskSizeRange() {
```
Comment (MathRPG) [inline src/ispd/arquivo/xml/utils/WrappedElement.java:39]: Could be renamed to "toUniformDistribution". This name is closer to what it actually does.

File: src/ispd/arquivo/xml/utils/WrappedElement.java [hunk 3]
```diff
@@ -13,6 +14,49 @@
...
+
+    /**
+     * Construct a {@link TwoStageUniform} from this instance, with the values
+     * acquired from {@link #minimum()}, {@link #maximum()},
+     * {@link #average()} and {@link #probability()}.
+     *
+     * @return {@link TwoStageUniform} with a minimum, maximum, average and
+     * probability values as the respective tags in this instance.
+     */
+    public TwoStageUniform toTaskSize() {
```
Comment (MathRPG) [inline src/ispd/arquivo/xml/utils/WrappedElement.java:51]: Again, method should be renamed.

File: src/ispd/gui/LoadConfigurationDialog.java [hunk 4]
```diff
@@ -1148,20 +1153,22 @@ private void onCancelClick(final ActionEvent evt) {
         this.setVisible(false);
     }
 
-    private static GerarCarga loadGeneratorFromTableRow(final List row) {
-        return new CargaForNode(
+    private static WorkloadGenerator loadGeneratorFromTableRow(final List row
+            , final Supplier<Integer> idSupplier) {
+        return new PerNodeWorkloadGenerator(
```
Comment (MathRPG) [inline src/ispd/gui/LoadConfigurationDialog.java:1158]: Should validate to make sure it's correct: the order here can easily be wrong unnoticeably.

File: src/ispd/motor/queueNetworks/CentrosDeServico.java [hunk 5]
```diff
@@ -453,7 +453,7 @@ public NoFila peekFilaMestre() {
     }
 
     public NoFila peekFilaComun() {
-        System.out.printf("getTipo() = %d \n", getTipo());
+        System.out.printf("getType() = %d \n", getTipo());
```
Comment (MathRPG) [inline src/ispd/motor/queueNetworks/CentrosDeServico.java:456]: Revert this, must've done by accident when changing the name of the method in the workload generator interface.

File: src/ispd/motor/random/TwoStageUniform.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+     * Construct an instance with all values set to 0.
+     */
+    public TwoStageUniform() {
+        this(0, 0, 0, 0);
+    }
+
+    /**
+     * Construct an instance with given {@code minimum} and {@code maximum}
+     * values, and with intervalSplit and firstIntervalProbability to
+     * represent a
```
Comment (MathRPG) [inline src/ispd/motor/random/TwoStageUniform.java:28]: Reformat.

File: src/ispd/motor/workload/impl/CollectionWorkloadGenerator.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+    /**
+     * @return the inner workload generator list.
+     */
+    public List<WorkloadGenerator> getList() {
+        return this.list;
+    }
+
+    /**
+     * Make a task list from the inner per-node load list, or an <b>empty</b>
```
Comment (MathRPG) [inline src/ispd/motor/workload/impl/CollectionWorkloadGenerator.java:47]: Make link to enum constant.

File: src/ispd/motor/workload/impl/PerNodeWorkloadGenerator.java [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+     *     <li>{@link #taskCount}</li>
+     *     <li>{@link #computation}'s interval maximum</li>
+     *     <li>{@link #computation}'s interval minimum</li>
+     *     <li>{@link #communication}'s interval maximum</li>
+     *     <li>{@link #communication}'s interval minimum</li>
+     * </ol>
+     *
+     * @return {@link Vector} with this instance's data.
+     * @see TwoStageUniform
+     */
```
Comment (MathRPG) [inline src/ispd/motor/workload/impl/PerNodeWorkloadGenerator.java:115]: Add link to method that interprets the generated vector, and vice-versa.

**Review by MathRPG — COMMENTED** [review]
Left comments with things that slipped the refactoring. They should be fixed before the pull request is merged.

File: src/ispd/arquivo/xml/utils/WrappedElement.java (hunk 1 above)
Comment (MathRPG) [reply src/ispd/arquivo/xml/utils/WrappedElement.java:26]: Instances created via this method are not used to generate values, only to gather the values in its fields. Still, in the future, a better concept should be created to encapsulate this.

File: src/ispd/gui/LoadConfigurationDialog.java [hunk 9]
```diff
@@ -1098,47 +1100,50 @@ private void setTipo(final int type) {
...
-                final int numTaref =
-                        (Integer) this.jSpinnerNumTarefas.getValue();
-                final int minComp =
-                        (Integer) this.jSpinnerMinComputacao.getValue();
-                final int maxComp =
-                        (Integer) this.jSpinnerMaxComputacao.getValue();
-                final int aveComp =
-                        (Integer) this.jSpinnerAverageComputacao.getValue();
+                final var minComp =
+                        (Double) this.jSpinnerMinComputacao.getValue();
```
Comment (MathRPG) [inline src/ispd/gui/LoadConfigurationDialog.java:1104]: Do these need to be wrapped casts?

File: src/ispd/arquivo/xml/models/builders/LoadBuilder.java [hunk 10]
```diff
@@ -57,68 +60,72 @@ public static Optional<? extends GerarCarga> build(final WrappedDocument doc) {
...
+                .map(el -> LoadBuilder.nodeLoadFromElement(el, idSupplier))
+                .map(WorkloadGenerator.class::cast)
                 .toList();
 
         if (nodeLoads.isEmpty()) {
             return Optional.empty();
         }
 
-        return Optional.of(new CargaList(nodeLoads, GerarCarga.FORNODE));
+        return Optional.of(new CollectionWorkloadGenerator(WorkloadGeneratorType.PER_NODE, nodeLoads
```
Comment (MathRPG) [inline src/ispd/arquivo/xml/models/builders/LoadBuilder.java:88]: Reformat

**Review by MathRPG — COMMENTED** [review]
Last pass through, LGTM.

**Review by ricardofares — APPROVED** [review]
* The modifications has been tested among different test cases and the results match the expected.

**Approved!**

**Pull request merged**